design-proposals/klip-8-interactive-queries.md

```diff
@@ -131,13 +131,130 @@ Additionally:
+
+Currently, a streaming query against a table in KSQL will only return rows from the table's changelog produced *after* the query begins (assuming a default start offset of `latest`).
+
+However, streaming queries returning a table's entire materialized state *then* followed by changes is more consistent with the behavior of PIT queries on a table. This behavior is also likely what users want when issuing a streaming query against a table, and it is currently not straightforward to achieve.
+
+To get KSQL's current behavior with a streaming query against a table, users would simply use an intuitive `WHERE` clause:
+
+```sql
+SELECT * FROM table WHERE ROWTIME >= NOW;
+```
+
+## Limitations
+
+The following are proposed intentional limitations of the initial implementation of PIT queries:
+
+**PIT queries on streams probably do not need to be initially supported.** The semantics and use cases around these are unclear and require further analysis. However, the proposed syntax leaves open the possibility of supporting PIT queries on streams without further modifications.
+
+**CTAS/CSAS defined by PIT queries do not initially need to be supported.** For example, the following CTAS would create a table, `t0`, and initially populate it with the result of the `SELECT` query:
+
+```sql
+CREATE TABLE t0 AS SELECT x, count(*) FROM t1 GROUP BY x;
+```
+
+**PIT queries against tables created directly over a topic (leaf tables) do not need to be initially supported.** Leaf tables are not materialized until another query reads from them. As a result, a PIT query against such a table would (potentially) not return the expected result. Rather than leave open the possibility of confusing query results in this scenario, we should simply disallow PIT queries against leaf tables until we decide to support them.
+
+---
+
+## Other considered approaches
+
+The following are other approaches that have been considered for identifying PIT versus streaming queries:
+
+### **Interpret any query against a table as PIT**
+
+The table-based approach would consider any SQL query against a KSQL table as a PIT query. Any query against a stream would then be a streaming query. The primary issue with this approach is that it complicates the capability of running a streaming query on a table, which is a highly valuable use case. Also, since PIT queries would be identified somewhat implicitly, this approach may not be particularly intuitive to users. It would also make it relatively easy to write a "correct" query that does the wrong thing.
```

Comment:
> are you referring to PIT on streams?

What is meant by implicit here is that PIT versus streaming behavior would not be expressed explicitly via syntax. So you couldn't look at a SQL query in isolation and know exactly how it will behave without having more context about the underlying relations and their types (streams or tables).

Reply:
> as we discussed above, as long as we support a `create stream table_change as select * from table`, we can achieve running streaming queries on a table, right? Is the additional sql state what you are concerned about for complexity?

I dug into this pretty deeply and technically speaking, it is *theoretically* possible but would require quite a bit of rework. But the semantics of "casting" a changelog to a stream become problematic due to the table's topic using compaction and [other subtle complications](https://github.com/confluentinc/ksql/pull/3117#discussion_r307279509). @big-andy-coates will be able to provide greater technical depth here.

Considering all of this, I don't see much upside to this approach.

Reply:
> Also, since PIT queries would be identified somewhat implicitly, this approach may not be particularly intuitive to users

I would argue the opposite. The present biggest source of confusion for new users, empirically speaking, is that `select * from table` returns an unbounded set. We see that question all the time.

Reply:
> I would argue the opposite. The present biggest source of confusion for new users, empirically speaking, is that select * from table returns an unbounded set. We see that question all the time.

I think that this specific example is pretty intuitive, but the complexity then arises from the context-dependent behavior in general. And with the proposed syntax, `SELECT * FROM table` will no longer return an unbounded set.

Reply:
I agree with the proposal (and thus @derekjn's POV) that we should avoid any implicit, context-dependent behavior. (Imagine you are reading a piece of Java code, but without diving into six other classes you can't tell what the line's doing that you are just looking at.)

To be more precise: I think implicit behavior and defaults are fine, but only if they are consistent (this is objective) and ideally intuitive (this is subjective). We should therefore *not* opt for a design where `SELECT * FROM myTable` gives you a PIT query whereas `SELECT * FROM myStream` gives you a streaming query.

Reply:
> We should therefore not opt for a design where SELECT * FROM myTable gives you a PIT query whereas SELECT * FROM myStream gives you a streaming query.

I don't understand your argument. Why would this be "inconsistent" or "context dependent behavior"? It's two different queries (because they are executed on different inputs, a STREAM vs a TABLE) and hence it seem ok (and actually quite intuitive---_subjectively_) that one terminate automatically (and has "snapshot semantics") while the other runs forever.